File: Python Iteration/Lecture.py
import sys
from abc import abstractmethod, ABC


class Lecture(ABC):

    def __init__(self, name, code, lecturer, credit, schedule, semester, grade=None):
        self.name = name
        self.code = code
        self.lecturer = lecturer
        self.credit = credit
        self.schedule = schedule
        self.semester = semester
        self.grade = grade
        self.passed = False
        self.registered_students = [[], [], [], [], [], [], [], []]

    @abstractmethod
    def register(self, student, semester):
        overlap = True
        studentsCourses = student.transcript[semester]
        days = self.parseDays(self.schedule)
        time = self.parseTime(self.schedule)
        for i in studentsCourses:
            days1 = self.parseDays(i.schedule)
            time1 = self.parseTime(i.schedule)
            if len(days) >= len(days1):
                overlap = self.checkOverlap(days1, days, time1, time)
            else:
                overlap = self.checkOverlap(days, days1, time, time1)
            if not overlap:
                print(f"Advisor didn't approve {self.name} because of a collision with { i.name} in schedule ", file=sys.stderr)
                return False
        return True

    def checkOverlap(self, smallSize, bigSize, time, time1):
        for i in range(len(smallSize)):
            for j in range(len(bigSize)):
                if smallSize[i] == bigSize[j]:
                    if self.isOverlapping([time[i], time1[j]]):
                        return False
        return True

    def parseDays(self, schedule):
        tokens = schedule.split()
        days = [tokens[i] for i in range(0, len(tokens), 2)]
        return days

    def parseTime(self, schedule):
        tokens = schedule.split()
        time = [tokens[i].split("-") for i in range(1, len(tokens), 2)]
        return time

    def isOverlapping(self, intervals):
        overlapping = [[x, y] for x in intervals for y in intervals if
                       x is not y and x[1] > y[0] > x[0] or x[0] == y[0] and x[1] == y[1] and x is not y]
        return not not overlapping

File: Java Iteration/src/main/java/model/Lecture.java
package model;

//import sun.plugin.javascript.navig.Array;



import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public abstract class Lecture {
    private String name;
    private String code;
    private Teacher lecturer;
    private int credit;
    private int ECTS;
    private boolean passed;
    private Lecture prerequisite;
    protected LinkedList<Student> registeredStudents;
    private String schedule;
    private String grade;



    public Lecture(String name, String code, Teacher lecturer, int credit, int ECTS, Lecture prerequisite,String schedule) {
        this.name = name;
        this.code = code;
        this.lecturer = lecturer;
        this.credit = credit;
        this.ECTS = ECTS;
        this.prerequisite = prerequisite;
        registeredStudents = new LinkedList<>();
        this.schedule = schedule;
    }

    @Override
    public String toString(){
        return "[Code:" + this.getCode() + " Name:" + this.getName() + " Lecturer:" +
                this.lecturer.getName() + " Credit:" + this.getCredit();
    }

    public int getCredit() {
        return credit;
    }

    public int getECTS() {
        return ECTS;
    }

    public boolean isPassed() {
        return passed;
    }

    public void setPassed(boolean passed) {
        this.passed = passed;
    }

    public Lecture getPrerequisite() {
        return prerequisite;
    }

    public void setPrerequisite(Lecture prerequisite) {
        this.prerequisite = prerequisite;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getCode() {
        return code;
    }

    public Teacher getLecturer() {
        return lecturer;
    }

    public String getSchedule() {
        return schedule;
    }

    public String getGrade() {
        return grade;
    }

    public void setGrade(String grade) {
        this.grade = grade;
    }

    public List<String> parseDays(String schedule){
        List<String> days= new ArrayList<String>();
        List<String> tokens = Arrays.asList(schedule.split(" "));

        for (int i=0; i<tokens.size() ;i+=2){
            days.add(tokens.get(i));
        }

        return days;
    }

    public List<String> parseTime(String schedule){
        List<String> time= new ArrayList<String>();
        List<String> tokens = Arrays.asList(schedule.split(" "));
        List<String> time1= new ArrayList<String>();

        for (int i=1; i<tokens.size() ;i+=2){
            time.add(tokens.get(i));
        }

        for (String s : time) {
            tokens = Arrays.asList(s.split("-"));

            for (String token : tokens) {
                time1.add(token);
            }
        }
        return time1;
    }
}
